### PR TITLE
maxima: add a subpkg maxima-ecl needed for sagemath

### PR DESCRIPTION
**Pull request by tornaria**
- Added a build option 'ecl' (enabled by default)
 - The ecl binary and library will be in a subpkg maxima-ecl so this
   should not affect current users of maxima
 - new patches:
   - a0d7a43...: build a FASL library for ECL (this is merged upstream)
   - handle-multiple-ldflags.patch: otherwise compilation with multiple options in LDFLAGS fails (taken from debian)
   - matrixexp.patch: fixes an error in matrix exponentiation (taken from debian, this originates in sagemath)

A weak point is that the library maxima.fas is installed in
/usr/lib/ecl-${ecl_version} which is where ecl looks for them, but this
doesn't seem right.

Maybe an alternative would be to have the ecl package ship a symlink at
/usr/lib/ecl pointing to the versioned directory, then have the
maxima-ecl package place its library in the non-versioned directory.

#### Testing the changes
- I tested the changes in this PR: **briefly**

To compile sagemath using system maxima, two things are necessary:
 - the maxima-ecl package that this PR introduces
 - the (yet unmerged) changes for sagemath in https://trac.sagemath.org/ticket/32867
Using both I compiled and doctested sagemath, and it seems to be working ok.

#### RFC
My doubts are:
- where to place the library `maxima.fas` as discussed above
- whether to have a build_option for ecl or just hardcode it as enabled
- whether to split the `maxima-ecl` package or place the two files it contains in the package `maxima` itself

### CODE REVIEW

**Comment by tornaria** [comment]
A few more comments:

a. this package is not marked nocross, but there's no way to build on cross since sbcl, clisp, emacs are nocross. OTOH emacs is not really required: I tried having emacs as a build option (disabled on cross) and using ecl to build maxima cross aarch64. However, this results in x86_64 binaries rather than aarch64 binaries. Ofc if we use target ecl it doesn't run and if I use host ecl it builds for host. We'd need a cross-ecl which we don't have, so we should flag the package as nocross, right?

b. there is a lline `vopt_conflict clisp sbcl`, but I don't think there's any conflict on those. In fact, I compiled maxima with the three backends (sbcl, clisp, ecl) and it builds and runs ok. You can run `maxima -l clisp` to use the clisp backend, etc. Note that when you compile with multiple backends sbcl is the default, that's why the default I set (sbcl + ecl) doesn't affect current users.

Note that sbcl is much more efficient by far. Running the testsuite with `echo 'run_testsuite();' | maxima` (and same with -l ecl, -l clisp) takes me:
 - sbcl: real	1m35.744s
 - ecl: real	8m54.797s
 - clisp: real	18m14.480s

This justifies the default choice of compiling with sbcl in the main package (maxima) and moving the ecl backend into a split subpkg which is used for sagemath (ofc it would be nice if sagemath was able to use maxima compiled with sbcl, but at this time it can't). I'm not sure what advantage clisp brings to the table.

**Comment by tornaria** [comment]
@jbgg @knusbaum: I'd appreciate if you have any comments about this PR which relates maxima and ecl.

TL;DR: we need maxima compiled with ecl as a dependency of sagemath, and that I'm not sure about where to place the library `maxima.fas` so that ecl can pick it up with `(require 'maxima)`. So far I placed it into `/usr/lib/ecl-21.2.1/` but that ties maxima pkg to the precise version of ecl it was compiled with. I'd rather place it in `/usr/lib/ecl/` but I don't know how to configure ecl so it looks in that directory too.

**Comment by dkwo** [comment]
If it does not affect maxima users and works with sage, then it looks good to me.

**Comment by knusbaum** [comment]
@tornaria 
I don't recall how xbps handles packages that require specific versions of another package.
As long as maxima-ecl gets rebuilt whenever ecl gets updated, it should be fine to include in `/usr/lib/ecl-21.2.1/`. I spent a little time in the ECL manual and on web search looking for a formal description of ecl's `require` implementation, but didn't find much. 

I *did* find that ecl fasls are basically shared objects, so they may need to be recompiled when ecl is updated anyway. I'm not sure and can't find any good info.

It's a shame, based on the [sagemath changes](https://trac.sagemath.org/ticket/32867) it looks like maxima needs to be `require`able rather than using any other package solution (`asdf` or somesuch)

**Comment by tornaria** [comment]
> @tornaria I don't recall how xbps handles packages that require specific versions of another package. As long as maxima-ecl gets rebuilt whenever ecl gets updated, it should be fine to include in `/usr/lib/ecl-21.2.1/`. I spent a little time in the ECL manual and on web search looking for a formal description of ecl's `require` implementation, but didn't find much.
> 
> I _did_ find that ecl fasls are basically shared objects, so they may need to be recompiled when ecl is updated anyway. I'm not sure and can't find any good info.

I was hoping that it doesn't need to be recompiled for minor version changes to ecl. In fact the binary which gets compiled in `/usr/lib/maxima/5.45.1/binary-ecl/maxima` requires `libecl.so.21.2` so I guess maxima will need recompilling (and xbps will enforce it) when the soname for libecl.so changes.

 > It's a shame, based on the [sagemath changes](https://trac.sagemath.org/ticket/32867) it looks like maxima needs to be `require`able rather than using any other package solution (`asdf` or somesuch)

This is where sagemath loads ecl and maxima.fas: [src/sage/interfaces/maxima_lib.py](https://github.com/sagemath/sage/blob/develop/src/sage/interfaces/maxima_lib.py#L99). As you can see, sagemath embeds ecl and runs lisp code to load maxima, etc.

I don't know much about common lisp. What would be a better (standard, portable) way to load `maxima.fas` using `asdf` instead of `require`? We could presumably try to change the initialization in sagemath. What would be the portable location for `maxima.fas` making this possible?

**Comment by tornaria** [comment]
In the old version there was an executable `maxima` and a library `maxima.fas` which where almost copies of each other.

In the new version I moved the common code to a shared library. The new package `maxima-ecl` contains:
 - `/usr/lib/maxima/5.45.1/binary-ecl/libmaxima-ecl.so`: shared library (13M)
 - `/usr/lib/maxima/5.45.1/binary-ecl/maxima`: binary for maxima -l ecl (20K)
 - `/usr/lib/maxima/5.45.1/binary-ecl/maxima.fas`: fasl for `(require 'maxima)` in ecl (20K)
 - `/usr/lib/ecl-21.2.1/maxima-fas` symlink to `maxima.fas` where ecl can find it

The way this is acomplished is placing all binaries (shared lib, executable, and fasl) in the same directory; the last two have `rpath=$ORIGIN` so they find the shared. An alternative is to place the shared library in `/usr/lib` to avoid the need for rpath.

Note that this change does _not_ affect current users of `maxima` pkg. All that changes is adding these four new files in a separate package `maxima-ecl`. Even installing this package doesn't affect current users of maxima since just running `maxima` in the cmdline will use the SBCL version by default. Only when doing `maxima -l ecl` will the ecl version be run. Alternatively, running ecl and doing `(require 'maxima)` loads maxima into ecl.

Given that the pkg `maxima-ecl` is just 10% size of the pkg `maxima`, it's also possible to just include those four files in the main pkg. That wouldn't be a problem and the only cost is this 10% extra space for users of `maxima` not interested in the ecl version of maxima.

This works well with sagemath and it would be really nice to have it available on the system, since maxima doesn't change as often as sagemath itself.

**Comment by tornaria** [comment]
IMO this is good for merging.

**Comment by tornaria** [comment]
@leahneukirchen What do you think? Could we merge this?
